### PR TITLE
Blokh/feat/persist file document and refactor worfklow service

### PR DESCRIPTION
**Pull request by Blokh**
### Description
Added the logic of persisting workflows using run.
refactored the implementation of workflow run

added the logic of finding, copying and persisting the file information from source to destination

### CODE REVIEW

File: services/workflows-service/src/providers/file/file.service.ts [hunk 1]
```diff
@@ -5,7 +5,9 @@ import {
   IStreamableFileProvider,
 } from './types';
 import * as tmp from 'tmp';
+import {Injectable} from "@nestjs/common";
```
Comment (alonp99) [inline services/workflows-service/src/providers/file/file.service.ts:8]: how it passed the lint check?

File: services/workflows-service/src/storage/get-file-storage-manager.ts [hunk 2]
```diff
@@ -15,15 +15,23 @@ export const S3StorageEnvSchema = z.object({
   AWS_S3_BUCKET_KEY: z.string(),
 });
 
-const __generateAwsConfig = (processEnv: NodeJS.ProcessEnv): S3ClientConfig => {
-  const { AWS_REGION, AWS_S3_BUCKET_KEY, AWS_S3_BUCKET_SECRET } =
-    S3StorageEnvSchema.parse(processEnv);
+export const generateAwsConfig = (processEnv: NodeJS.ProcessEnv, prefixConfigName?: string): S3ClientConfig => {
+  const { AWS_REGION, AWS_S3_BUCKET_KEY, AWS_S3_BUCKET_SECRET } = S3StorageEnvSchema.parse(processEnv);
```
Comment (alonp99) [inline services/workflows-service/src/storage/get-file-storage-manager.ts:19]: I wouldve expect to have S3-specific stuff in some s3 adapter, so someone can easily duplicate it and add another provider

File: services/workflows-service/src/storage/storage.controller.internal.ts [hunk 3]
```diff
@@ -13,6 +13,7 @@ import {
   fetchDefaultBucketName,
   manageFileByProvider,
 } from '@/storage/get-file-storage-manager';
+import console from 'console';
```
Comment (alonp99) [inline services/workflows-service/src/storage/storage.controller.internal.ts:16]: remove

File: services/workflows-service/src/workflow/workflow.service.ts [hunk 4]
```diff
@@ -19,12 +19,24 @@ import { BusinessRepository } from '@/business/business.repository';
...
+import { TFileServiceProvider, TRemoteFileConfig, TS3BucketConfig } from '@/providers/file/types';
+import { z } from 'zod';
+import { HttpFileService } from '@/providers/file/file-provider/http-file.service';
+import { LocalFileService } from '@/providers/file/file-provider/local-file.service';
+import { AwsS3FileService } from '@/providers/file/file-provider/aws-s3-file.service';
+import { StorageService } from '@/storage/storage.service';
+import { FileService } from '@/providers/file/file.service';
+import * as process from 'process';
+import { generateAwsConfig } from '@/storage/get-file-storage-manager';
+import * as crypto from 'crypto';
```
Comment (alonp99) [inline services/workflows-service/src/workflow/workflow.service.ts:32]: add line breaks between import and the rest of the code, try to have types in a designated file

File: services/workflows-service/src/workflow/workflow.service.ts [hunk 5]
```diff
@@ -19,12 +19,24 @@ import { BusinessRepository } from '@/business/business.repository';
...
+type TEntityId = string;
 const ajv = new Ajv({
   strict: false,
 });
+import { env } from '@/env';
```
Comment (alonp99) [inline services/workflows-service/src/workflow/workflow.service.ts:37]: hoist this to all the imports

File: services/workflows-service/src/workflow/workflow.service.ts [hunk 6]
```diff
@@ -19,12 +19,24 @@ import { BusinessRepository } from '@/business/business.repository';
...
 addFormats(ajv, { formats: ['email', 'uri'] });
-
+type TDefaultSchemaDocumentPage = DefaultContextSchema['documents'][number]['pages'][number];
```
Comment (alonp99) [inline services/workflows-service/src/workflow/workflow.service.ts:39]: move to /schemas

File: services/workflows-service/src/workflow/workflow.service.ts [hunk 7]
```diff
@@ -381,6 +337,148 @@ export class WorkflowService {
...
+    return { ...context, documents: documentsWithPersistedImages };
+  }
+
+  private async __persistDocumentPagesFiles(
+    document: DefaultContextSchema['documents'][number],
+    entityId: string,
+  ) {
+    return await Promise.all(
+      document.pages.map(async documentPage => {
+        const id = `${document?.category}-${document?.type}-${document?.issuer?.country}`;
```
Comment (alonp99) [inline services/workflows-service/src/workflow/workflow.service.ts:360]: used?

File: services/workflows-service/src/workflow/workflow.service.ts [hunk 8]
```diff
@@ -381,6 +337,148 @@ export class WorkflowService {
...
+
+        const documentContext =
+          `${document?.category}-${document?.type}-${document?.issuer?.country}`.toLowerCase();
+        const documentName = `${entityId}/${documentContext}_${crypto.randomUUID()}.${
```
Comment (alonp99) [inline services/workflows-service/src/workflow/workflow.service.ts:364]: does the entity folder worked eventually?

File: services/workflows-service/src/workflow/workflow.service.ts [hunk 9]
```diff
@@ -381,6 +337,148 @@ export class WorkflowService {
...
+          documentPage.type
+        }`;
+
+        const { fromServiceProvider, fromRemoteFileConfig } =
+          this.__fetchFromServiceProviders(documentPage);
+        const { toServiceProvider, toRemoteFileConfig, remoteFileName } =
+          this.__fetchToServiceProviders(documentName);
+
+        const remoteFilePath = await this.fileService.copyFileFromSourceToSource(
```
Comment (alonp99) [inline services/workflows-service/src/workflow/workflow.service.ts:373]: I think source to source is confusing, there is a source and destination

File: services/workflows-service/src/workflow/workflow.service.ts [hunk 10]
```diff
@@ -381,6 +337,148 @@ export class WorkflowService {
...
+      data: {
+        correlationId: entity.id,
+        ...(context.entity.data as object),
+      } as Prisma.BusinessCreateInput,
+    });
+
+    return id;
+  }
+
+  private async __tryToFetchExistingEntityIc(entity: {
```
Comment (alonp99) [inline services/workflows-service/src/workflow/workflow.service.ts:452]: Typo

File: services/workflows-service/src/workflow/workflow.service.ts [hunk 11]
```diff
@@ -381,6 +337,148 @@ export class WorkflowService {
...
+        const remoteFilePath = await this.fileService.copyFileFromSourceToSource(
+          fromServiceProvider,
+          fromRemoteFileConfig,
+          toServiceProvider,
+          toRemoteFileConfig,
+        );
+        const fileNameInBucket =
+          typeof remoteFilePath != 'string' ? remoteFilePath.fileNameInBucket : undefined;
+        const userId = entityId;
+        const ballerineFileId = await this.storageService.createFileLink({
```
Comment (alonp99) [inline services/workflows-service/src/workflow/workflow.service.ts:382]: check if ballerineFileId exists, if it does - skip

File: services/workflows-service/src/storage/get-file-storage-manager.ts (hunk 2 above)
Comment (Blokh) [reply services/workflows-service/src/storage/get-file-storage-manager.ts:19]: I need suggestions regarding how should I approach this issue

File: services/workflows-service/src/workflow/workflow.service.ts (hunk 11 above)
Comment (Blokh) [reply services/workflows-service/src/workflow/workflow.service.ts]: resolved

**Review by alonp99 — APPROVED** [review]

**Pull request merged**